File: Python/Program/Garage.py
from Car import Car 
from Motorcycle import Motorcycle
from ParkingLot import ParkingLot

class Garage:
    __namaGarasi = ""
    __luas = 0
    __vehicles = []
    
    def __init__(self,namaGarasi,luas):
        self.__namaGarasi = namaGarasi
        self.__luas = luas
        self.__vehicles = []
        
    def setNamaGarasi(self, namaGarasi):
        self.__namaGarasi = namaGarasi
        
    def getNamaGarasi(self):
        return self.__namaGarasi
    
    def setLuasGarasi(self, luasGarasi):
        self.__luas = luasGarasi
        
    def getNamaGarasi(self):
        return self.__luas
    
    def setVehicles(self, vehicle):
        self.__vehicles = vehicle
        
    def getVehicles(self):
        return self.__vehicles
    
    def moveToParkingLot(self, vehicle, parkingLot):
        parkingLot.park_vehicle(vehicle) 
        
    def add_vehicle(self, vehicle, parkingLot):
        if len(self.__vehicles) < 2:
            self.__vehicles.append(vehicle) 
            print("Berhasil menambahkan kendaraan dengan Plat", vehicle.getNoPlat(), "ke garasi")
        else:
            print("Garasi Penuh, Kendaraan dengan Plat", vehicle.getNoPlat(), "dipindahkan ke Praking Lot")
            self.moveToParkingLot(vehicle, parkingLot)
        
    def display_vehicle(self):
        print(">>>> List Kendaraan di", self.__namaGarasi + ":")
        for vehicle in self.__vehicles:
            print("Plat Nomor :", vehicle.getNoPlat())
            print("Merk :", vehicle.getMerk())
            print("Tahun Produksi :", vehicle.getTahunProduksi())
            print("Warna :", vehicle.getWarna())
            if isinstance(vehicle, Car):
                print("Banyak Kursi :", vehicle.getJumlahKursi())
                print("Banyak Pintu :", vehicle.getJumlahPintu())
                print("Banyak Roda :", vehicle.getJumlahRoda())
                print("Tipe :", vehicle.getTipe())
            elif isinstance(vehicle, Motorcycle):
                print("Tipe :", vehicle.getTipe())
                print("Kapasitas Tangki :", vehicle.getKapasitasTangki())
            print()
            
    

File: CPP/Program/Garage.cpp
#pragma once

#include <string>
#include <iostream>
#include <vector>
#include "Car.cpp"
#include "Motorcycle.cpp"
#include "ParkingLot.cpp"

using namespace std;

class Garage {
private:
    string namaGarasi;
    int luas;
    vector<Vehicle*> vehicles;

public:
    Garage(string namaGarasi, int luas) : namaGarasi(namaGarasi), luas(luas) {}

    void moveToParkingLot(Vehicle* vehicle, ParkingLot* parkingLot) {
        parkingLot->park_vehicle(vehicle);
    }

    void add_vehicle(Vehicle* vehicle, ParkingLot* parkingLot) {
        if (vehicles.size() < 2) {
            vehicles.push_back(vehicle);
            cout << "Berhasil menambahkan kendaraan dengan Plat " << vehicle->getNoPlat() << " ke garasi\n";
        } else {
            cout << "Garasi Penuh, Kendaraan dengan Plat " << vehicle->getNoPlat() << " dipindahkan ke Parking Lot\n";
            moveToParkingLot(vehicle, parkingLot);
        }
    }

    void display_vehicle() {
        cout << ">>>> List Kendaraan di " << namaGarasi << ":\n";
        for (Vehicle* vehicle : vehicles) {
            cout << "Plat Nomor : " << vehicle->getNoPlat() << "\n";
            cout << "Merk : " << vehicle->getMerk() << "\n";
            cout << "Tahun Produksi : " << vehicle->getTahunProduksi() << "\n";
            cout << "Warna : " << vehicle->getWarna() << "\n";
            if (dynamic_cast<Car*>(vehicle)) {
                Car* car = dynamic_cast<Car*>(vehicle);
                cout << "Banyak Kursi : " << car->getJumlahKursi() << "\n";
                cout << "Banyak Pintu : " << car->getJumlahPintu() << "\n";
                cout << "Banyak Roda : " << car->getJumlahRoda() << "\n";
                cout << "Tipe : " << car->getTipe() << "\n";
            } else if (dynamic_cast<Motorcycle*>(vehicle)) {
                Motorcycle* motor = dynamic_cast<Motorcycle*>(vehicle);
                cout << "Tipe : " << motor->getTipe() << "\n";
                cout << "Kapasitas Tangki : " << motor->getKapasitasTangki() << "\n";
            }
            cout << "\n";
        }
    }
};